test/src/test/java/jenkins/model/JenkinsLogRecordsTest.java

```diff
@@ -37,14 +37,19 @@ private static void _logRecordsArePresentOnController(JenkinsRule r) throws Thro
             containsInRelativeOrder("Completed initialization", "Started initialization"));
         VersionNumber javaVersion = new VersionNumber(System.getProperty("java.specification.version"));
         if (javaVersion.isOlderThan(new VersionNumber("9")) || javaVersion.isNewerThanOrEqualTo(new VersionNumber("17"))) { // TODO https://github.com/jenkinsci/jenkins-test-harness/issues/359
-            LogRecord lr = new LogRecord(Level.INFO, "collect me");
-            Logger.getLogger(Jenkins.class.getName()).log(lr);
-            WeakReference<LogRecord> ref = new WeakReference<>(lr);
-            lr = null;
+            Object x = new Object() {
+                @Override
+                public String toString() {
+                    return "collect me";
+                }
+            };
+            Logger.getLogger(Jenkins.class.getName()).log(Level.INFO, "formatting {0}", x);
+            WeakReference<Object> ref = new WeakReference<>(x);
+            x = null;
             MemoryAssert.assertGC(ref, true);
-            assertThat("Records collected",
+            assertThat("Record parameters formatted before collection",
                 logRecords.stream().map(LogRecord::getMessage).collect(Collectors.toList()),
-                hasItem("<discarded>"));
+                hasItem("formatting collect me"));
```

Comment:
The `ProxyException` logic is not covered by tests and is therefore subject to regression. Adding test coverage for this logic reduces this risk of regression:

```suggestion
            Logger logger = Logger.getLogger(Jenkins.class.getName());
            LogRecord lr = new LogRecord(Level.INFO, "formatting {0}");
            Object params[] = {x};
            lr.setLoggerName(logger.getName());
            lr.setParameters(params);
            lr.setThrown(new Throwable());
            logger.log(lr);
            WeakReference<Object> ref = new WeakReference<>(x);
            x = null;
            MemoryAssert.assertGC(ref, true);
            assertThat("Record parameters formatted before collection",
                logRecords.stream().map(LogRecord::getMessage).collect(Collectors.toList()),
                hasItem("formatting collect me"));
            assertThat("Large objects that might be held by the stack trace not held",
                 logRecords.stream()
                      .filter(record -> record.getMessage().contains("formatting collect me"))
                      .map(LogRecord::getThrown)
                      .findFirst()
                      .orElseThrow(() -> new AssertionError("not found")),
                 instanceOf(ProxyException.class));
```

(Plus corresponding imports.)

Reply:
Yes, it is probably wise to do this, though I would rather assert the _goal_ (that an object present in a field of the `Throwable` is not held) than the _mechanism_ (`ProxyException`).

Reply:
Even a mere assertion of the mechanism prevents the mechanism from accidentally being undone by a future change, which is better than nothing. Obviously an assertion of the goal would be even better.